Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs

```diff
@@ -131,52 +133,57 @@ handler merchant sReq = do
       merchantId = merchant.id
   result <- getDistanceAndDuration merchantId fromLocationLatLong toLocationLatLong sReq.routeDistance sReq.routeDuration
   logDebug $ "distance: " <> show result.distance
-  mbSpecialLocation <- QSpecialLocation.findSpecialLocationByLatLong fromLocationLatLong
   sessiontoken <- generateGUIDText
   fromLocation <- buildSearchReqLocation merchantId sessiontoken sReq.pickupAddress sReq.customerLanguage sReq.pickupLocation
   toLocation <- buildSearchReqLocation merchantId sessiontoken sReq.dropAddrress sReq.customerLanguage sReq.dropLocation
 
+  mbSpecialLocationTag <- FareProduct.getSpecialLocationTag merchantId fromLocationLatLong toLocationLatLong
+  allFarePolicies <- getFarePolicyForVariants merchantId fromLocationLatLong toLocationLatLong
+  let farePolicies = selectFarePolicy result.distance allFarePolicies
   (quotes, mbEstimateInfos) <-
-    if isJust mbSpecialLocation
+    if isJust mbSpecialLocationTag
       then do
         whenJustM
           (QSearchRequestSpecialZone.findByMsgIdAndBapIdAndBppId sReq.messageId sReq.bapId merchantId)
           (\_ -> throwError $ InvalidRequest "Duplicate Search request")
         searchRequestSpecialZone <- buildSearchRequestSpecialZone sReq merchantId fromLocation toLocation result.distance result.duration
         Esq.runTransaction $ do
           QSearchRequestSpecialZone.create searchRequestSpecialZone
-        now <- getCurrentTime
-        listOfSpecialZoneQuotes <- do
-          allFarePolicies <- FarePolicyS.findAllByMerchantId merchant.id
-          let farePolicies = selectFarePolicy result.distance allFarePolicies
-          let listOfVehicleVariants = listVehicleVariantHelper farePolicies
-          for listOfVehicleVariants $ \farePolicy -> do
-            fareParams <-
-              calculateFareParameters
-                CalculateFareParametersParams
-                  { farePolicy = farePolicy,
-                    distance = result.distance,
-                    rideTime = sReq.pickupTime,
-                    waitingTime = Nothing,
-                    driverSelectedFare = Nothing,
-                    customerExtraFee = Nothing
-                  }
-            buildSpecialZoneQuote
-              searchRequestSpecialZone
-              fareParams
-              merchant.id
-              result.distance
-              farePolicy.vehicleVariant
-              result.duration
-        Esq.runTransaction $
-          for_ listOfSpecialZoneQuotes QQuoteSpecialZone.create
-        return (Just (mkQuoteInfo fromLocation toLocation now <$> listOfSpecialZoneQuotes), Nothing)
+        flowType <- getFlowType merchantId fromLocationLatLong toLocationLatLong
```

Comment:
`getFareProductForVariants` query is repeated in `getFlowType` and `getFarePolicyForVariants`, can we query once and use at same value at both the places ?

Reply:
Yes but getFlowType is used only in one place, we dont want to compute it all other places where FarePolicy is required. One thing is that I always calculate FareProduct outside the query and then pass it but then whereever FarePolicy is required first i will have to always query FareProduct and pass it to get FarePolicy, I rather put all that logic for linking with FareProduct at one place.

Reply:
Is there some better way to do it, I dont want to compute FareProduct outside and pass it as an argument to `getFarePolicyForVariants`

Reply:
Yes, Cool I handled this case, I have put flow in Full Fare Policy to avoid DB call.